Build failure: pkgsCross.aarch64-multiplatform.python312Packages.psycopg

## Steps To Reproduce

Steps to reproduce the behavior:

1. Copy this Nix flake

   ```
   {
     inputs = {
       nixpkgs.url = "github:nixos/nixpkgs?ref=nixos-unstable";
     };
     outputs = { self, nixpkgs }: {
       nixosConfigurations.test = nixpkgs.lib.nixosSystem {
         system = "aarch64-linux";
         modules = [ {
           system.stateVersion = "24.05";
           boot.loader.grub.devices = [ "/dev/sda" ];
           fileSystems."/" = {
             device = "/dev/sda1";
             fsType = "ext4";
           };
           services.paperless = {
             enable = true;
           };
         } ];
       };
       packages.aarch64-linux = {
         paperless-ngx = nixpkgs.legacyPackages.aarch64-linux.paperless-ngx;
         psycopg = nixpkgs.legacyPackages.aarch64-linux.python312Packages.psycopg;
       };
     };
   }
   ```

2. Run `nix build .#nixosConfigurations.test.config.system.build.toplevel`

3. After about 2 hours (I am cross-compiling _to_ aarch64-linux), I see the following error:

   ```
   error: builder for '/nix/store/qqn9pscl8021bah21s4ph31km6c0m7yd-python3.12-psycopg-3.2.3.drv' failed with exit code 1;

          ...trimming the XPASS, XFAIL, and PASS lines...

          > FAILED tests/pool/test_pool_common_async.py::test_setup_no_timeout[asyncio-AsyncNullConnectionPool] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool_common_async.py::test_setup_no_timeout[asyncio-AsyncConnectionPool] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool_async.py::test_stats_connect[asyncio] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool_async.py::test_refill_on_check[asyncio] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool_async.py::test_connect_check_timeout[asyncio] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool_async.py::test_reconnect_after_grow_failed[asyncio] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool_null_async.py::test_stats_connect[asyncio] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool.py::test_stats_connect - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool.py::test_reconnect_after_grow_failed - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool.py::test_refill_on_check - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool.py::test_connect_check_timeout - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool.py::test_uniform_use - assert {1, 2, 3} == {2}
          > FAILED tests/test_concurrency_async.py::test_eintr[ITIMER_REAL-SIGALRM] - AssertionError: script terminated with SIGALRM
          > FAILED tests/test_concurrency_async.py::test_ctrl_c - AssertionError: process didn't start?
          > FAILED tests/pool/test_pool_common.py::test_setup_no_timeout[NullConnectionPool] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool_common.py::test_setup_no_timeout[ConnectionPool] - ValueError: the proxy didn't start listening in time
          > FAILED tests/pool/test_pool_null.py::test_stats_connect - ValueError: the proxy didn't start listening in time
          > FAILED tests/test_concurrency.py::test_eintr[ITIMER_REAL-SIGALRM] - AssertionError: script terminated with SIGALRM
          > FAILED tests/test_concurrency.py::test_ctrl_c - AssertionError: process didn't start?
          > = 19 failed, 4919 passed, 197 skipped, 454 deselected, 37 xfailed, 2 xpassed in 2873.81s (0:47:53) =
          For full logs, run 'nix log /nix/store/qqn9pscl8021bah21s4ph31km6c0m7yd-python3.12-psycopg-3.2.3.drv'.
   error: 1 dependencies of derivation '/nix/store/5k9siwszm3hj4r9ajdrj6lgg2065538q-paperless-ngx-2.13.5.drv' failed to build
   error: 1 dependencies of derivation '/nix/store/g0db28n49pc6xiqg68ybl0kdxagdzisl-unit-paperless-web.service.drv' failed to build
   error: 1 dependencies of derivation '/nix/store/phjs36rg03alc4idb7lydgjn3mi9n3xm-system-units.drv' failed to build
   error: 1 dependencies of derivation '/nix/store/gpvi9g9paljdqp2cfhwi24sjr9kqkkjf-etc.drv' failed to build
   error: 1 dependencies of derivation '/nix/store/q0sbyx27snsj963x8rn74x2i7fh108gl-nixos-system-remote-25.05.20241213.3566ab7.drv' failed to build
   ```

## Build log

<!-- insert build log in code block in collapsable section -->

[Build Log](https://dpaste.org/bhFGc), too big for GitHub issue

## Additional context

1. Not sure if this matters, but I am on an `x86_64-linux` host, cross-compiling for `aarch64-linux`.

1. The odd thing is that `nix build .#psycopg2 --system aarch64-linux` (and `paperless-ngx`) build correctly. I can't isolate the failure except by building a whole NixOS Configuration.

2. I am unable to disable tests on psycopg, probably because I don't understand how to override it correctly.

   ```
   paperless.package = pkgs.paperless-ngx.override (paperless-ngx-old-attrs: {
     python3 = paperless-ngx-old-attrs.python3.override {
        packageOverrides = self: pypkgs-old: {
         psycopg = pypkgs-old.psycopg.overridePythonAttrs {
           pytestCheckPhase = ''true'';
           doCheck = false;
         };
       };
     };
   });
   ``` 

3. Why doesn't `pytestCheckPhase` have `-x` (stop after first failure)? Should it?

4. Too my untrained eye, all of the broken should probably be skipped because they would conflict with the Nix sandbox.

   - `test_setup_no_timeout`, `test_reconnect_after_grow_failed`, `test_setup_no_timeout`: create warnings like the following:

     ```
     WARNING  psycopg.pool:pool.py:637 error connecting in 'pool-225': connection failed: connection to server at "127.0.0.1", port 54975 failed: Connection refused
             Is the server running on that host and accepting TCP/IP connections?
     ```

## Metadata

```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.56, NixOS, 24.11 (Vicuna), 24.11.20241014.a3c0b3b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.25.0pre20240920_ca3fc169`
 - nixpkgs: `/home/sam/.nix-defexpr/channels/nixpkgs`
```

## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@SuperSandro2000 (last non-treewide committer)

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/psycopg/default.nix
+++ b/pkgs/development/python-modules/psycopg/default.nix
@@ -173,7 +173,7 @@ buildPythonPackage rec {
       pytestCheckHook
       postgresql
     ]
-    ++ lib.optional (stdenv.hostPlatform.isLinux) postgresqlTestHook
+    ++ lib.optional stdenv.hostPlatform.isLinux postgresqlTestHook
     ++ optional-dependencies.c
     ++ optional-dependencies.pool;
 
@@ -187,7 +187,7 @@ buildPythonPackage rec {
     ''
       cd ..
     ''
-    + lib.optionalString (stdenv.hostPlatform.isLinux) ''
+    + lib.optionalString stdenv.hostPlatform.isLinux ''
       export PSYCOPG_TEST_DSN="host=/build/run/postgresql user=$PGUSER"
     '';
 
@@ -219,7 +219,7 @@ buildPythonPackage rec {
   ];
 
   postCheck = ''
-    cd ${pname}
+    cd psycopg
   '';
 
   passthru = {

--- a/pkgs/development/python-modules/psycopg/default.nix
+++ b/pkgs/development/python-modules/psycopg/default.nix
@@ -121,10 +121,13 @@ buildPythonPackage rec {
 
   disabled = pythonOlder "3.7";
 
-  outputs = [
-    "out"
-    "doc"
-  ];
+  outputs =
+    [
+      "out"
+    ]
+    ++ lib.optionals (stdenv.hostPlatform == stdenv.buildPlatform) [
+      "doc"
+    ];
 
   sphinxRoot = "../docs";
 
@@ -141,13 +144,18 @@ buildPythonPackage rec {
     cd psycopg
   '';
 
-  nativeBuildInputs = [
-    furo
-    setuptools
-    shapely
-    sphinx-autodoc-typehints
-    sphinxHook
-  ];
+  nativeBuildInputs =
+    [
+      furo
+      setuptools
+      shapely
+    ]
+    # building the docs fails with the following error when cross compiling
+    #  AttributeError: module 'psycopg_c.pq' has no attribute '__impl__'
+    ++ lib.optionals (stdenv.hostPlatform == stdenv.buildPlatform) [
+      sphinx-autodoc-typehints
+      sphinxHook
+    ];
 
   propagatedBuildInputs = [
     psycopg-c

